More ParseException handling for numeric dimensions

jon-wei opened the pull request:
Fixes #4879 

This is a different approach to #4509 

The patch also adjusts InputRowSerde to serialize dimension values with the types specified in the ingestion schema, instead of serializing all values as Strings. ParseException handling for dimensions has been added to InputRowSerde.toBytes();

Regarding null/unparseable value handling:
- Unparseable strings in numeric dimensions will cause the column value to be replaced with a default zero when reportParseExceptions is false (this null->zero conversion should be removed when the #5278 series of patches is merged). When reportParseExceptions is true, unparseable strings in a numeric dimension will throw ParseException.
- The empty string in a numeric dimension is an unparseable value, it is not equivalent to null
- An actual null value (if a row explictly has a null value for a numeric dimension, or if the row is missing that dimension altogether) will still be converted to 0 for now. This should also change after #5278 allows Druid to store nulls for numeric columns

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -210,9 +211,17 @@ public boolean run()
-                                   .findCounter(HadoopDruidIndexerConfig.IndexJobCounters.INVALID_ROW_COUNTER);
-      jobStats.setInvalidRowCount(invalidRowCount.getValue());
+      Counters counters = job.getCounters();
+      if (counters == null) {
```
Why'd you need to add these? Why would the counters not be set?

jon-wei replied:
It happened on a task failure I was testing the ParseExceptions, didn't look further into why the job counters weren't set

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -252,12 +261,15 @@ private static IncrementalIndex makeIncrementalIndex(
     return newIndex;
   }
 
+
```
Too many newlines here.

jon-wei replied:
fixed

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java:
```diff
@@ -49,7 +57,166 @@
+    }
+  }
+
+  public static final byte[] toBytes(
```
Please add `@Nullable` and a javadoc explaining what it means when this returns null (probably it means there was a parse exception and `reportParseExceptions` is false? Is it possible for this to return null if `reportParseExceptions` is true?).

gianm replied:
I think this might never return null if you take one of the other suggestions (skip column instead of skip whole row on parse error).

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java:
```diff
@@ -49,7 +57,166 @@
+    Map<String, IndexSerdeTypeHelper> typeHelperMap = Maps.newHashMap();
+    for (DimensionSchema dimensionSchema : dimensionsSpec.getDimensions()) {
+      IndexSerdeTypeHelper typeHelper;
+      switch (dimensionSchema.getValueType()) {
```
Replace the switch with `VALUE_TYPE_HELPER_ARRAY[dimensionSchema.getValueType().ordinal()]` ?

jon-wei replied:
I kept the switch statement but dropped the array, it wasn't being used anymore after removing the type ordinals from the serialized form

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java:
```diff
@@ -49,7 +57,166 @@
+  {
+    ValueType getType();
+
+    byte[] serialize(Object value);
```
`void serialize(DataOutput out, Object value)` would be better - avoid needless byte[] allocations.

jon-wei replied:
changed to  `void serialize(ByteArrayDataOutput out, Object value)`

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java:
```diff
@@ -63,9 +230,25 @@
+            byte[] value = typeHelper.serialize(row.getRaw(dim));
+
+            writeString(dim, out);
+            WritableUtils.writeVInt(out, typeHelper.getType().ordinal());
```
Writing the type ordinal isn't necessary since the reader should know the type from the name of the dim.

jon-wei replied:
dropped the type ordinal

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java:
```diff
@@ -49,7 +57,166 @@
+
+    byte[] serialize(Object value);
+
+    T deserialize(byte[] bytes);
```
`T deserialize(DataInput in)` is better for similar reasons.

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -381,11 +409,16 @@ private void flushIndexToContextAndClose(BytesWritable key, IncrementalIndex ind
+        byte[] serializedRow = InputRowSerde.toBytes(typeHelperMap, inputRow, combiningAggs, true);
+
+        if (serializedRow != null) {
+          // reportParseExceptions is true as any unparseable data is already handled by the mapper.
```
This comment is attached to the wrong line. It should go with toBytes.

jon-wei replied:
Moved comment

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java:
```diff
@@ -63,9 +230,25 @@
+            if (reportParseExceptions) {
+              throw pe;
+            } else {
+              // discard the row if there was a parse error in a dimension
```
I think it would be better to just use a "default" value for this column like zero. It's more in line with what aggregators do (see the below log message "Encountered parse error, skipping aggregator"). That way, at least we retain _something_ from the row (namely: all other columns).

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -299,9 +316,13 @@ protected void innerMap(
+                                  InputRowSerde.toBytes(typeHelperMap, inputRow, aggregators, reportParseExceptions);
+
+      if (serializedInputRow == null) {
+        return;
```
This should increment INVALID_ROW_COUNTER and log a debug message. However, this comment may be moot if you take the suggestion to skip just the column and not the entire row.

jon-wei replied:
Added a counter increment and debug message for now

gianm commented on processing/src/main/java/io/druid/segment/LongDimensionIndexer.java:
```diff
@@ -40,7 +40,7 @@ public Long processRowValsToUnsortedEncodedKeyComponent(Object dimValues)
     }
 
-    return DimensionHandlerUtils.convertObjectToLong(dimValues);
+    return DimensionHandlerUtils.convertObjectToLong(dimValues, true);
```
If we get an unparseable string value here, and reportParseExceptions is false, I think it'd be better to replace it with zero than to skip the entire row. That way we aren't throwing away as much data.

If reportParseExceptions is true then yes, we should throw an exception out.

(+ similar comment for other numeric types)

jon-wei commented:
@gianm I addressed your comments aside from the ones regarding rejecting entire rows on parse errors vs. replacing unparseable values with a default value.

I went with the row rejection approach since I think ingesting a partially correct row is introducing more "corruption" into the data than just throwing the row away, on some datasets zero could be a significant value for a numeric dimension and adding the unparseable rows would lead to confusing results. 

I saw that the metrics handle parse exceptions by skipping the aggregator, but I also lean towards throwing entire rows away in that case (maybe you have a set of metrics that are somewhat related/correlated, like maybe you have a metric for an total count + a metric for a related flow rate, if there's a parse error in one of the values I think it's arguably weird to ingest the row)

I don't feel super strongly about this though, and it's not absolutely clear to me that "partially correct" rows are better/worse than missing rows, so I'll defer to your judgment or the community's if others wish to comment on this.

gianm commented:
@jon-wei my feeling is that if you want total integrity then that is what `reportParseExceptions` is for. You will know for sure if you get a bad value because ingestion will just stop and wait for you to fix it. If you don't set reportParseExceptions then I think it makes sense to ingest whatever columns we can ingest. It means that for queries that don't touch this column then they will be more correct (since all other columns are fine). Queries that do touch this column will be worse off, but there are more other-columns than this-columns, so I think the balance is in favor of zeroing out the column and keeping the rest of the row.

jon-wei commented:
@gianm That reasoning sounds good to me, I've updated the patch to replace unparseable numeric dimension values with zeros when reportParseExceptions is false

jihoonson commented:
I'm reviewing this PR.

jihoonson commented on processing/src/main/java/io/druid/query/filter/DoubleValueMatcherColumnSelectorStrategy.java:
```diff
@@ -31,7 +31,7 @@
   public ValueMatcher makeValueMatcher(final BaseDoubleColumnValueSelector selector, final String value)
   {
-    final Double matchVal = DimensionHandlerUtils.convertObjectToDouble(value);
+    final Double matchVal = DimensionHandlerUtils.convertObjectToDouble(value, false);
```
nit: Probably better to add a method `DimensionHandlerUtils.convertObjectToDouble(value)` which internally calls `DimensionHandlerUtils.convertObjectToDouble(value, false)` for convenience.

jon-wei replied:
Added a method without the parseException boolean

jon-wei commented:
@jihoonson 

> What do you think about supporting configurable default values for unparseable values?

I think that'd be a useful feature.

I think it would be better to implement default values for unparseable values in a separate PR (this one is mainly to fix an NPE bug), and get #5278 in before that patch to avoid introducing more potential conflicts there since that's a big patch that's been open for a while.

jihoonson commented:
@jon-wei ok. That sounds good to me.
